pkg/controllermanager/controller/shoot/shoot_maintenance_control.go

```diff
@@ -224,3 +221,101 @@ func hasMaintainNowAnnotation(shoot *gardenv1beta1.Shoot) bool {
+		return canUpdateToLatestMachineImageVersion(shootCurrentImage, machineImageFromCloudProfile)
+	}
+
+	return false, nil, nil
+}
+
+func shootMachineImageExistsInCloudProfile(shoot *gardenv1beta1.Shoot, machineImageFromCloudProfile gardenv1beta1.MachineImage) (bool, error) {
+	cloudProvider, err := helper.GetShootCloudProvider(shoot)
+	if err != nil {
+		return false, err
+	}
+	shootMachineImage := helper.GetMachineImageFromShoot(cloudProvider, shoot)
+	exists, _ := helper.ShootMachineImageVersionExists(machineImageFromCloudProfile, shootMachineImage)
+	return exists, nil
+}
+
+// canUpdateToLatestMachineImageVersion checks if the Shoot's machine image can be updated to the latest version
+func canUpdateToLatestMachineImageVersion(shootCurrentImage *gardenv1beta1.ShootMachineImage, machineImageFromCloudProfile gardenv1beta1.MachineImage) (bool, *gardenv1beta1.ShootMachineImage, error) {
+	versionUpdatePossible, updateMachineImage, err := VersionUpdatePossible(shootCurrentImage, machineImageFromCloudProfile)
+	if err != nil {
+		return false, nil, err
+	}
+	if versionUpdatePossible {
+		return true, updateMachineImage, nil
+	}
+	return false, nil, nil
+}
+
+// updateToLatestMachineImageVersion returns the latest machine image and requiring an image update
+func updateToLatestMachineImageVersion(machineImage gardenv1beta1.MachineImage) (bool, *gardenv1beta1.ShootMachineImage, error) {
+	_, latestMachineImage, err := helper.GetShootMachineImageFromLatestMachineImageVersion(machineImage)
+	if err != nil {
+		return false, nil, fmt.Errorf("failed to determine latest machine image in cloud profile: %s", err.Error())
+	}
+	return true, &latestMachineImage, nil
+}
+
+// VersionUpdatePossible checks if the shoot's current image can be updated to a higher semantic version
+func VersionUpdatePossible(shootCurrentImage *gardenv1beta1.ShootMachineImage, machineImageFromCloudProfile gardenv1beta1.MachineImage) (bool, *gardenv1beta1.ShootMachineImage, error) {
+	latestSemVerVersion, latestMachineImage, err := helper.GetShootMachineImageFromLatestMachineImageVersion(machineImageFromCloudProfile)
+	if err != nil {
+		return false, nil, fmt.Errorf("failed to determine latest machine image in cloud profile: %s", err.Error())
+	}
+
+	shootImageSemVerVersion, err := semver.NewVersion(shootCurrentImage.Version)
+	if err != nil {
+		return false, nil, fmt.Errorf("error while parsing machine image version of shoot (invalid): %s", err.Error())
+	}
+
+	if latestSemVerVersion.GreaterThan(shootImageSemVerVersion) {
+		return true, &latestMachineImage, nil
+	}
+	return false, nil, nil
+}
+
+// ForceUpdateRequired checks if the shoots current machine image has to be forcefully updated
+func ForceUpdateRequired(shootCurrentImage *gardenv1beta1.ShootMachineImage, imageCloudProvider gardenv1beta1.MachineImage) bool {
+	for _, image := range imageCloudProvider.Versions {
+		if shootCurrentImage.Version != image.Version {
+			continue
+		}
+
+		if image.ExpirationDate == nil {
```

Comment:
```suggestion
		return ExpirationDateExpired(image.ExpirationDate) {
```

It doesn't make sense to continue the loop - you have found the machine image in the cloud profile with the correct version at this point. Also, to simplify the code, I would directly return the result of the function `ExpirationDateExpired` here, where `ExpirationDateExpired` is:

```golang
// ExpirationDateExpired returns if now is equal or after the given expirationDate
func ExpirationDateExpired(timestamp *metav1.Time) bool {
	if timestamp == nil {
		return false
	}
	return time.Now().UTC().After(timestamp.Time) || time.Now().UTC().Equal(timestamp.Time)
}
```